### PR TITLE
opencpn: add darwin support

### PR DESCRIPTION
**Pull request by wegank**
###### Description of changes

As in the title.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/opencpn/default.nix [hunk 1]
```diff
@@ -48,9 +51,32 @@ stdenv.mkDerivation rec {
     hash = "sha256-sNZYf/2gtjRrrGPuazVnKTgcuIQpKPazhexqlK21T4g=";
   };
 
-  nativeBuildInputs = [ cmake lsb-release pkg-config ];
+  patches = [
+    (fetchpatch {
+      url = "https://patch-diff.githubusercontent.com/raw/OpenCPN/OpenCPN/pull/2473.patch";
```
Comment (lovesegfault) [inline pkgs/applications/misc/opencpn/default.nix:56]: Since the PR was merged, please use the URL from the commit to master:

https://github.com/OpenCPN/OpenCPN/commit/30fa16850ba97d3df0622273947e3e3975b8e6c0.patch

File: pkgs/applications/misc/opencpn/default.nix [hunk 2]
```diff
@@ -48,9 +51,32 @@ stdenv.mkDerivation rec {
...
+      sha256 = "sha256-Sb4FE9QJA5kMJi52/x1Az6rMTS3WSURPx4QAhcv2j9E=";
+    })
+  ];
+
+  postPatch = ''
+    substituteInPlace cmake/TargetSetup.cmake \
+      --replace '"sw_vers" "-productVersion"' '"echo" "1"'
+    sed -i '/fixup_bundle/d' CMakeLists.txt
```
Comment (lovesegfault) [inline pkgs/applications/misc/opencpn/default.nix:64]: Why is this needed unconditionally now?

File: pkgs/development/libraries/libunarr/default.nix [hunk 3]
```diff
@@ -16,11 +16,14 @@ stdenv.mkDerivation rec {
     substituteInPlace pkg-config.pc.cmake \
       --replace '$'{prefix}/@CMAKE_INSTALL_LIBDIR@ @CMAKE_INSTALL_FULL_LIBDIR@ \
       --replace '$'{prefix}/@CMAKE_INSTALL_INCLUDEDIR@ @CMAKE_INSTALL_FULL_INCLUDEDIR@
+  '' + lib.optionalString stdenv.isDarwin ''
+    substituteInPlace CMakeLists.txt \
+      --replace '-Wl,--no-undefined -Wl,--as-needed' '-Wl,-undefined,error'
```
Comment (lovesegfault) [inline pkgs/development/libraries/libunarr/default.nix:21]: Add a comment justifying why this is needed.

File: pkgs/applications/misc/opencpn/default.nix (hunk 2 above)
Comment (wegank) [reply pkgs/applications/misc/opencpn/default.nix:64]: Since these commands are darwin-only and won't affect Linux builds.

File: pkgs/development/libraries/libunarr/default.nix (hunk 3 above)
Comment (wegank) [reply pkgs/development/libraries/libunarr/default.nix]: Done.

File: pkgs/applications/misc/opencpn/default.nix (hunk 2 above)
Comment (lovesegfault) [reply pkgs/applications/misc/opencpn/default.nix:64]: Then move them to the portion that's darwin-only, IMO

Comment (wegank) [reply pkgs/applications/misc/opencpn/default.nix:64]: Done.
